datafusion/core/src/datasource/file_format/parquet.rs

```diff
@@ -719,55 +717,300 @@ impl DataSink for ParquetSink {
+                    .create_all_async_arrow_writers(
+                        num_partitions,
+                        parquet_props,
+                        object_store.clone(),
+                    )
+                    .await?;
+                // TODO parallelize individual parquet serialization when already outputting multiple parquet files
+                // e.g. if outputting 2 parquet files on a system with 32 threads, spawn 16 tasks for each individual
+                // file to be serialized.
+                row_count = output_multiple_parquet_files(writers, data).await?;
+            }
+            true => {
+                if !allow_single_file_parallelism || data.len() <= 1 {
+                    let mut writer = self
+                        .create_all_async_arrow_writers(
+                            num_partitions,
+                            parquet_props,
+                            object_store.clone(),
+                        )
+                        .await?
+                        .remove(0);
+                    for data_stream in data.iter_mut() {
                         while let Some(batch) = data_stream.next().await.transpose()? {
-                            cnt += batch.num_rows();
+                            row_count += batch.num_rows();
                             writer.write(&batch).await?;
                         }
-                        writer.close().await?;
-                        Ok(cnt)
-                    });
+                    }
+
+                    writer.close().await?;
+                } else {
+                    let object_store_writer = self
+                        .create_object_store_writers(1, object_store)
+                        .await?
+                        .remove(0);
+                    row_count = output_single_parquet_file_parallelized(
+                        object_store_writer,
+                        data,
+                        self.config.output_schema.clone(),
+                        parquet_props,
+                    )
+                    .await?;
                 }
-                while let Some(result) = join_set.join_next().await {
-                    match result {
-                        Ok(res) => {
-                            row_count += res?;
-                        } // propagate DataFusion error
-                        Err(e) => {
-                            if e.is_panic() {
-                                std::panic::resume_unwind(e.into_panic());
-                            } else {
-                                unreachable!();
+            }
+        }
+
+        Ok(row_count as u64)
+    }
+}
+
+/// This is the return type when joining subtasks which are serializing parquet files
+/// into memory buffers. The first part of the tuple is the parquet bytes and the
+/// second is how many rows were written into the file.
+type ParquetFileSerializedResult = Result<(Vec<u8>, usize), DataFusionError>;
```

Comment:
Writing to disk as an option would be a cool feature. It might be too slow to spill to disk vs just writing on a single thread, but certainly worth testing that empirically. 

I think the best possible solution would consume the sub parquet files incrementally from memory as they are produced, rather than buffering the entire file.

Reply:
I agree that incremental consumption of the files (and applying backpressure if the final stage can't keep up) is likely the the best tradeoff.

Reply:
filed https://github.com/apache/arrow-datafusion/issues/7591